### PR TITLE
Change CRIU RNG to be non-blocking, mixing in SHA1PRNG data

### PR DESCRIPTION
**Pull request by alon-sh**
Signed-off-by: Alon Shalev Housfater <alonsh@ca.ibm.com>

JDK17 version of https://github.com/ibmruntimes/openj9-openjdk-jdk11/pull/535/

### CODE REVIEW

**Comment by alon-sh** [comment]
FYI @tajila

**Comment by tajila** [comment]
@JasonFengJ9 please take a look

**Comment by tajila** [comment]
@alon-sh we will also need this in jdknext

File: closed/src/java.base/share/classes/openj9/internal/criu/NativePRNG.java [hunk 1]
```diff
@@ -219,30 +237,49 @@ private byte[] implGenerateSeed(int numBytes) {
...
+                    data[i] ^= rawData[i];
                 }
+            } catch (IOException e) {
+                throw new ProviderException("nextBytes() failed", e);
+            }
+        }
+
+        private void clearRNGState() {
+            if (mixRandom != null) {
+                mixRandom.clearState();
```
[the remark this answers is not in the document]
Comment (alon-sh) [reply closed/src/java.base/share/classes/openj9/internal/criu/NativePRNG.java:280]: mixRandom doesn't get re-used. The clear state is invoked at a post-checkpoint hook, during restore - the providers are changed and CRIUSEC provider is removed from providers list.

File: closed/src/java.base/share/classes/openj9/internal/criu/NativePRNG.java [hunk 2]
```diff
@@ -219,30 +237,49 @@ private byte[] implGenerateSeed(int numBytes) {
...
+                   seedOut.write(seed);
+               } catch (IOException e) {
+                   // ignored, on Mac OS X, /dev/urandom can be opened
+                   // for write, but actual write is not permitted
+               }
+           }
+           getMixRandom().engineSetSeed(seed);
+        }
+
+        private synchronized void implNextBytes(byte[] data) {
```
[the remark this answers is not in the document]
Comment (alon-sh) [reply closed/src/java.base/share/classes/openj9/internal/criu/NativePRNG.java:264]: updated

**Comment by alon-sh** [comment]
@JasonFengJ9 I updated the code - thanks

**Comment by JasonFengJ9** [comment]
jenkins test sanity zlinux,x86-64_linux_criu jdk17

**Comment by JasonFengJ9** [comment]
```
Testing: Create CRIU checkpoint image and restore three times - testMillisDelayBeforeCheckpointDone
Test start time: 2022/06/15 18:32:18 Eastern Standard Time
Running command: bash /home/jenkins/workspace/Test_openjdk17_j9_sanity.functional_x86-64_linux_criu_Personal_testList_1/aqa-tests/TKG/../../jvmtest/functional/cmdLineTests/criu/criuScript.sh /home/jenkins/workspace/Test_openjdk17_j9_sanity.functional_x86-64_linux_criu_Personal_testList_1/aqa-tests/TKG/../../jvmtest/functional/cmdLineTests/criu /home/jenkins/workspace/Test_openjdk17_j9_sanity.functional_x86-64_linux_criu_Personal_testList_1/openjdkbinary/j2sdk-image/bin/java " -Xint -XX:+CRIURestoreNonPortableMode " "org.openj9.criu.TimeChangeTest testMillisDelayAfterCheckpointDone" ThreeCheckpoints
Time spent starting: 5 milliseconds
Time spent executing: 18868 milliseconds
Test result: FAILED
Output from test:
 [OUT] start running script
 [OUT] Start test name: testMillisDelayAfterCheckpointDone
 [OUT] Before starting testMillisDelayAfterCheckpointDone: System.currentTimeMillis: 1655332338295, Wed Jun 15 18:32:18 EDT 2022, Thread's name: main
 [OUT] taskMillisDelayAfterCheckpointDone: System.currentTimeMillis: 1655332343647, Wed Jun 15 18:32:23 EDT 2022, Thread's name: Timer-0
 [OUT] FAILED: expected MILLIS_DELAY_AFTERCHECKPOINTDONE 5000 ms, but the actual elapsed time is: 3196 ms
 [OUT] End testMillisDelayAfterCheckpointDone: System.currentTimeMillis: 1655332352908, Wed Jun 15 18:32:32 EDT 2022, Thread's name: main
 [OUT] finished script
 [ERR] /home/jenkins/workspace/Test_openjdk17_j9_sanity.functional_x86-64_linux_criu_Personal_testList_1/aqa-tests/TKG/../../jvmtest/functional/cmdLineTests/criu/criuScript.sh: line 26:  3448 Killed                  $2 -XX:+EnableCRIUSupport $3 -cp "$1/criu.jar" $4 $5 > testOutput 2>&1
 [ERR]   3448: Error (criu/files-reg.c:1994): File home/jenkins/workspace/Test_openjdk17_j9_sanity.functional_x86-64_linux_criu_Personal_testList_1/aqa-tests/TKG/output_16553274205740/cmdLineTester_criu_nonPortableRestore_1/testOutput has bad size 555 (expect 195)
 [ERR]   3448: Error (criu/files.c:1200): Unable to open fd=1 id=0x2d
 [ERR] Error (criu/cr-restore.c:1476): 3448 exited, status=1
 [ERR] Error (criu/cr-restore.c:2447): Restoring FAILED.
 [ERR]   3448: Error (criu/files-reg.c:1994): File home/jenkins/workspace/Test_openjdk17_j9_sanity.functional_x86-64_linux_criu_Personal_testList_1/aqa-tests/TKG/output_16553274205740/cmdLineTester_criu_nonPortableRestore_1/testOutput has bad size 555 (expect 195)
 [ERR]   3448: Error (criu/files.c:1200): Unable to open fd=1 id=0x2d
 [ERR] Error (criu/cr-restore.c:1476): 3448 exited, status=1
 [ERR] Error (criu/cr-restore.c:2447): Restoring FAILED.
>> Success condition was found: [Output match: Killed]
>> Required condition was found: [Output match: Start test name: testMillisDelay]
>> Required condition was not found: [Output match: PASSED: expected MILLIS_DELAY]
>> Failure condition was not found: [Output match: CRIU is not enabled]
>> Failure condition was not found: [Output match: Operation not permitted]
>> Failure condition was found: [Output match: FAILED: expected MILLIS_DELAY]
>> Required condition was found: [Output match: Error]

---TEST RESULTS---
Number of PASSED tests: 6 out of 7
Number of FAILED tests: 1 out of 7

---SUMMARY OF FAILED TESTS---
Create CRIU checkpoint image and restore three times - testMillisDelayBeforeCheckpointDone
-----------------------------


cmdLineTester_criu_nonPortableRestore_1_FAILED
```
This is because https://github.com/ibmruntimes/openj9-openjdk-jdk17/pull/110 hasn't been merged yet.

**Review by JasonFengJ9 — APPROVED** [review]

**Pull request merged**